### ISSUE TITLE
ICE: "cat_expr Errd"

### ISSUE THREAD

**Issue by cactorium** [rust-lang/rust#24504]
The compiler appears to crash when fed this invalid syntax.

I tried this code:

```
fn main() {
    let a = [];
}
```
I expected to see a syntax error, but instead I got this trace:

```
e0309.rs:2:13: 2:15 error: internal compiler error: cat_expr Errd
e0309.rs:2     let a = [];
                       ^~
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/beta-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:130

```

Backtrace is this:
```
stack backtrace:
   1:     0x7efbf74743c8 - sys::backtrace::write::h179e141147f9ee14IJC
   2:     0x7efbf749f770 - panicking::on_panic::h6aa569d33389f0cfsXI
   3:     0x7efbf73d74a3 - rt::unwind::begin_unwind_inner::h082f07869f4b9731ACI
   4:     0x7efbf479d04d - rt::unwind::begin_unwind::h13428428822164950735
   5:     0x7efbf479cff3 - diagnostic::SpanHandler::span_bug::he82cb6427e6a50a8jnB
   6:     0x7efbf503eb63 - session::Session::span_bug::h460607360e8bf7cc5Xq
   7:     0x7efbf6100d0d - check::regionck::visit_expr::hfc7cf3e25c050a92PJd
   8:     0x7efbf610183a - check::regionck::visit_local::h3c0931e616562244UGd
   9:     0x7efbf60fbd05 - check::regionck::Rcx<'a, 'tcx>::visit_fn_body::hd4da730c0cf65ac9amd
  10:     0x7efbf6196b91 - check::check_bare_fn::hb43e600383cf2ee2UQn
  11:     0x7efbf6190674 - check::check_item::h77d9b4d4612f23cfF9n
  12:     0x7efbf6268a96 - check_crate::closure.36300
  13:     0x7efbf6262e73 - check_crate::h4b0848761f18a2bcbwC
  14:     0x7efbf7af35cd - driver::phase_3_run_analysis_passes::hf7cad4098b28ceb1gGa
  15:     0x7efbf7ad7425 - driver::compile_input::h7c62c2870ff901f6Qba
  16:     0x7efbf7b8ea25 - run_compiler::hc815b49ce638c28eV4b
  17:     0x7efbf7b8c33d - boxed::F.FnBox<A>::call_box::h4290333406520961126
  18:     0x7efbf7b8b879 - rt::unwind::try::try_fn::h13999576404798567721
  19:     0x7efbf751bc68 - rust_try_inner
  20:     0x7efbf751bc55 - rust_try
  21:     0x7efbf7b8bb4b - boxed::F.FnBox<A>::call_box::h14489442628055152240
  22:     0x7efbf748a781 - sys::thread::create::thread_start::h02b7da32b4d7a4aaExH
  23:     0x7efbf127b373 - start_thread
  24:     0x7efbf704a27c - __clone
  25:                0x0 - <unknown>

```

**Comment by jakub-** [comment]
Thanks for your report! This issue is a duplicate of https://github.com/rust-lang/rust/issues/22897 so I'm going to close it.
